### ISSUE TITLE
Incorrect dot result with BinaryBackend

### ISSUE THREAD

**Issue by h3xby** [elixir-nx/nx#612]
```elixir
a = Nx.tensor([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
b = Nx.tensor([[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]])
c = Nx.dot(a, [0, 1], b, [1, 0])
IO.inspect(c)
```
```
#Nx.Tensor<
  f32
  55.0
>
```

With pytorch(and torchx backend):

```python
import torch

a = torch.tensor([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
b = torch.tensor([[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]])
c = torch.tensordot(a, b, dims=([0, 1],[1, 0]))
print(c)
```

```
tensor(50.)
```

Looks like BinaryBackend ignores tensor shapes and computes as if they are vectors: `0*0 + 1*1 + 2*2 + 3*3 + 4*4 + 5*5 = 55`

Also because of this [at least one test](https://github.com/elixir-nx/nx/blob/efed8861da20a256e6d328f61cd7c01bf4090620/nx/test/nx_test.exs#L1118) is incorrect

**Comment by josevalim** [comment]
Good catch! If you want to send a PR< it would be very welcome, otherwise I will take a look at it in the coming days.

**Comment by floatn** [comment]
Looks like it's the `aggregate_axes` function's fault. It's basically indifferent to input axis order:
https://github.com/elixir-nx/nx/blob/efed8861da20a256e6d328f61cd7c01bf4090620/nx/lib/nx/binary_backend.ex#L2145
```elixir
t = Nx.iota({4, 4, 4})
Nx.dot(t, [2, 0], t, [1, 0])
```
for example, returns the correct value because the axis values are resorted in the same fashion.
